### PR TITLE
Fix #119

### LINKED ISSUE #119: [ ALIGNMENT ]
## Summary

BSM implementation makes it awkward to match go-sdk and bsvjs Bsm implementations.

## Motivation

It's deprecated so we didn't focus on making it good but now it's raised its head we should implement this under the hood.

## Description
BSM.sign() should return the same base64 you'd expect from older libraries. We should be careful not to break ts-paymail in the process.

## Additional References
Here's the solution
```js
// bsv
import bsv from 'bsv'
import Message from 'bsv/message'
// @bsv/sdk
import { BSM, PrivateKey, Utils, Hash, BigNumber } from "@bsv/sdk"
const { toArray, Writer } = Utils

// sign with both libraries and compare signatures
const wif = "L211enC224G1kV8pyyq7bjVd9SxZebnRYEzzM3i7ZHCc1c5E7dQu"

const bsvSdkPk = PrivateKey.fromWif(wif)
const bsvPk = bsv.PrivateKey.fromWIF(wif)
const msgStr = "hello world"

const prefix = 'Bitcoin Signed Message:\n'
export const magicHash = (messageBuf: number[]): number[] => {
  const bw = new Writer()
  bw.writeVarIntNum(prefix.length)
  bw.write(toArray(prefix, 'utf8'))
  bw.writeVarIntNum(messageBuf.length)
  bw.write(messageBuf)
  const buf = bw.toArray()
  const hashBuf = Hash.hash256(buf)
  return hashBuf
}

const bsvSdkSig = BSM.sign(toArray(msgStr), bsvSdkPk)
const h = new BigNumber(magicHash(toArray(msgStr, 'utf8')))
const r = bsvSdkSig.CalculateRecoveryFactor(bsvSdkPk.toPublicKey(), h)
const sig = bsvSdkSig.toCompact(r, true, 'base64')
const bsvSig = new Message(msgStr).sign(bsvPk)

console.log("Sigs:", { sig, bsvSig })

// PRINTS
// 'Sigs:' {
//   sig: 'H4T8Asr0WkC6wYfBESR6pCAfECtdsPM4fwiSQ2qndFi8dVtv/mrOFaySx9xQE7j24ugoJ4iGnsRwAC8QwaoHOXk=',
//   bsvSig: 'H4T8Asr0WkC6wYfBESR6pCAfECtdsPM4fwiSQ2qndFi8dVtv/mrOFaySx9xQE7j24ugoJ4iGnsRwAC8QwaoHOXk='
// }
```

### PR DESCRIPTION
**Pull request by ty-everett**
## Description of Changes

Added a new base64 mode to the BSM implementation based on the code provided by @sirdeggen 

## Linked Issues / Tickets

Closes #119 

## Testing Procedure

Added a test vector that employs the new base64 mode.

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have updated `CHANGELOG.md` with my changes
- [x] I have run `npm run doc` and `npm run lint` one final time before requesting a review
- [x] I have run `npm version patch` so that my changes will trigger a new version to be released when they are merged

### CODE REVIEW

**Comment by ty-everett** [comment]
Note: I'm also adding a fix for #109 in this PR, reducing overhead of having lots of PRs.

File: src/compat/BSM.ts [hunk 1]
```diff
@@ -30,11 +30,18 @@ export const magicHash = (messageBuf: number[]): number[] => {
  * @deprecated Replaced by BRC-77 which employs BRC-42 key derivation and BRC-43 invoice numbers for enhanced security and privacy.
  * @param message The message to be signed as a number array.
  * @param privateKey The private key used for signing the message.
- * @returns The signature object.
+ * @param mode The mode of operation. When "raw", a Signature object is returned. When "base64", the BSM format signature is returned. Default: "raw".
+ * @returns The signature object when in raw mode, or the BSM base64 string when in base64 mode.
  */
-export const sign = (message: number[], privateKey: PrivateKey): Signature => {
+export const sign = (message: number[], privateKey: PrivateKey, mode: 'raw' | 'base64' = 'raw'): Signature | string => {
```
Comment (sirdeggen) [inline src/compat/BSM.ts:36]: I think the expectation here is that it returns base64 so we should have that be the default.

**Review by sirdeggen — CHANGES_REQUESTED** [review]

**Comment by sirdeggen** [comment]
Also please use address rather than public key in the verification process as David requests in the issue discussion.

**Comment by ty-everett** [comment]
Implementing a full `BSM.verify()` with Address support in the TS-SDK presents significant challenges that I do not currently have the capacity to handle.

Firstly, this task would necessitate implementing a new Address representation or accepting a string type, which introduces ambiguity since it could represent either a PublicKey or an Address. Moreover, converting a Signature type into a public key requires that all Signature types include a "recovery" parameter, which would need to be stored.

While supporting `sig.toPublicKey()` for all signatures might seem like a convenient way for new developers to provide the necessary input for the `ECDSA.verify()` function, it would not genuinely ensure that verification is taking place. The signature could originate from any source, and this could lead to unintended misuse.

It is important to recognize that adding a feature to compute the public key from a digital signature could introduce significant security risks, especially in a library intended for use by developers who may not be deeply familiar with cryptographic principles. While this might be mathematically feasible, it is not advisable. Based on my experience, I have seen too many instances of production code where such features were used improperly, leading to potential vulnerabilities.

Furthermore, deriving a public key from an address is inherently impossible due to the hash function involved. Using a signature for this purpose, as would be required here, exceeds the boundaries of what is reasonable in the name of compatibility.

Given these challenges, my proposed solution is for people to utilize a different library for the purpose of verifying a BSM signature from an Address, accepting any potential performance trade-offs as a natural consequence of avoiding a flawed approach within `@bsv/sdk`. At this time, I do not think it would be wise for me to delve back into the complexities of elliptic curve mathematics for what I consider a problematic task. I also believe I may not be the most qualified person to handle any implementation that involves the complexities of recovery parameters.

However, I am willing to flip the default as requested and I've committed the current work. Additionally, due to Amaury encoding, I have extracted the `minimallyEncode` code into a utility function.

I'd like to merge this code, as there are a few good pieces here. If it's decided that `BSM.verify()` with Address support is **absolutely required**, a new PR can be raised by someone who is able to implement it effectively.

**Pull request merged**